xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-io/src/main/java/org/xwiki/annotation/io/internal/DefaultIOService.java

```diff
@@ -200,12 +207,13 @@ public Collection<Annotation> getAnnotations(String target) throws IOServiceExce
                 return Collections.emptySet();
             }
             for (BaseObject object : objects) {
-                // if it's not on the required target, ignore it
-                if (object == null || !localTargetId.equals(object.getStringValue(Annotation.TARGET_FIELD))) {
-                    continue;
-                }
                 // use the object number as annotation id
-                result.add(loadAnnotationFromObject(object, deprecatedContext));
+                if (object != null) {
+                    String targetField = object.getStringValue(Annotation.TARGET_FIELD);
+                    if (Objects.equals(localTargetId, targetField) || Objects.equals("", targetField)) {
```

Comment:
I'm failing to remember exactly, but if I'm correct the check with empty string is for backward compatibility no? It might worth a comment here.